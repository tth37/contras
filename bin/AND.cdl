import Not from "Not.cdl"
@Head
NAME And
INPUT a, b
OUTPUT out
SYMBOL NAND NAND1
SYMBOL Not Not1
@Body
NAND1(a = a, b = b) => (out = nab)
Not1(a = nab) => (out = ab)
@End
OUTPUT out = ab