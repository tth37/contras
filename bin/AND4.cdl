import And from  "And.cdl"
@Head
NAME And4
INPUT a[3:0]
OUTPUT out
SYMBOL And And1, And2, And3
@Body
And1(a = a[0], b = a[1]) => (out = a01)
And2(a = a[2], b = a[3]) => (out = a23) 
And3(a = a01, b = a23) => (out = a0123)
@End
OUTPUT out = a0123