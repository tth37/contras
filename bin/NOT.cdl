@Head
NAME Not
INPUT a
OUTPUT out
SYMBOL NAND Nand1

@Body
Nand1(a = a, b = a) => (out = res)

@End
OUTPUT out = res

# .\contras.exe .\Not.cdl .\test\Not.in .\test\Not.out