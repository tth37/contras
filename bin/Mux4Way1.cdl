import AND4 from "AND4.cdl"
import NOT from "NOT.cdl"
import AND from "AND.cdl"
@Head
NAME Mux4Way1
INPUT a, b
INPUT D0, D1, D2, D3
OUTPUT W, NW
SYMBOL AND AND1, AND2, AND3, AND4
SYMBOL AND4 AND41
SYMBOL NAND NAND1, NAND2, NAND3, NAND4
SYMBOL NOT NOT1, NOT2, NOT3

@Body
NOT1(a=a)=>(out=na)
NOT2(a=b)=>(out=nb)
AND1(a=na, b=nb)=>(out=nanb)
AND2(a=a, b=nb)=>(out=anb)
AND3(a=na, b=b)=>(out=nab)
AND4(a=a, b=b)=>(out=ab)
NAND1(a=nanb, b=D0)=>(out=d0)
NAND2(a=anb, b=D1)=>(out=d1)
NAND3(a=nab, b=D2)=>(out=d2)
NAND4(a=ab, b=D3)=>(out=d3)
AND4(a=d0, b=d1, c=d2, d=d3)=>(out=nresult)
NOT3(a=nresult)=>(out=result)

@End
OUTPUT W=result, NW=nresult
