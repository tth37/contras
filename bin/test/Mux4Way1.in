@Case
a=0
b=1
D0=1
D1=0
D2=1
D3=0
W=?
NW=?
@End