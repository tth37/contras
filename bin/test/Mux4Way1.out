@Case
 a[0] = 010100101010101
 b[0] = 101010101000101
D0[0] = 110101010101010
D1[0] = 000000000000000
D2[0] = 111111100000000
D3[0] = 010010101010110
 W[0] = 101011111101110
NW[0] = 010100000010001
